File: boarding_sim/strategies.py
"""Boarding policy definitions, independent of physical boarding engines."""

from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass

from .models import Passenger
from .prng import RNG


def _seat_group(seat: str) -> int:
    return 0 if seat in {"A", "F"} else 1 if seat in {"B", "E"} else 2


def _side(seat: str) -> int:
    return 0 if seat in {"A", "B", "C"} else 1


def _zone_back_to_front(row: int) -> int:
    return (30 - row) // 5


@dataclass(frozen=True)
class Strategy:
    id: str
    name: str
    access_recommended: str
    prep_cohorts: int
    cohort: Callable[[Passenger], int]
    rank: Callable[[Passenger, float], float]
    door: Callable[[Passenger], str]
    preserve_seat_door: bool = False


STRATEGIES: dict[str, Strategy] = {
    "random_front": Strategy(
        "random_front", "Random · front door", "bridge", 1,
        lambda _p: 0, lambda _p, random_key: random_key, lambda _p: "front",
    ),
    "split_half_two_door": Strategy(
        "split_half_two_door", "Rows 1–15 front · 16–30 rear", "bus", 2,
        lambda p: 0 if p.row <= 15 else 1,
        lambda p, random_key: (0 if p.row <= 15 else 1) * 1000 + random_key,
        lambda p: "front" if p.row <= 15 else "rear", True,
    ),
    "wilma": Strategy(
        "wilma", "A/F → B/E → C/D", "bridge", 3,
        lambda p: _seat_group(p.seat),
        lambda p, random_key: _seat_group(p.seat) * 1000 + random_key,
        lambda _p: "front",
    ),
    "back_to_front_zones": Strategy(
        "back_to_front_zones", "Back-to-front · 5-row zones", "bridge", 6,
        lambda p: _zone_back_to_front(p.row),
        lambda p, random_key: _zone_back_to_front(p.row) * 1000 + random_key,
        lambda _p: "front",
    ),
    "wilma_zones": Strategy(
        "wilma_zones", "Outside-in + back-to-front zones", "bridge", 18,
        lambda p: _seat_group(p.seat) * 6 + _zone_back_to_front(p.row),
        lambda p, random_key: (_seat_group(p.seat) * 6 + _zone_back_to_front(p.row)) * 1000 + random_key,
        lambda _p: "front",
    ),
    "steffen_companion": Strategy(
        "steffen_companion", "Steffen-style · companion compatible", "bridge", 12,
        lambda p: _seat_group(p.seat) * 4 + _side(p.seat) * 2 + (p.row % 2),
        lambda p, _random_key: (_seat_group(p.seat) * 4 + _side(p.seat) * 2 + (p.row % 2)) * 1000 + (31 - p.row),
        lambda _p: "front",
    ),
    "split_wilma_two_door": Strategy(
        "split_wilma_two_door", "Split doors + A/F → B/E → C/D", "bus", 6,
        lambda p: (0 if p.row <= 15 else 1) * 3 + _seat_group(p.seat),
        lambda p, random_key: ((0 if p.row <= 15 else 1) * 3 + _seat_group(p.seat)) * 1000 + random_key,
        lambda p: "front" if p.row <= 15 else "rear", True,
    ),
}


def strategy_by_id(strategy_id: str) -> Strategy:
    return STRATEGIES[strategy_id]


def strategy_complexity(strategy: Strategy) -> float:
    if strategy.prep_cohorts <= 1:
        return 0.0
    return math.log2(strategy.prep_cohorts) / math.log2(18)


def strategy_catalog() -> list[dict[str, object]]:
    return [
        {
            "id": strategy.id,
            "name": strategy.name,
            "recommendedAccess": strategy.access_recommended,
            "preparationCohorts": strategy.prep_cohorts,
        }
        for strategy in STRATEGIES.values()
    ]


def apply_companion_compatibility(
    passengers: list[Passenger], strategy: Strategy, rng: RNG
) -> list[Passenger]:
    families: dict[int, list[Passenger]] = {}
    for passenger in passengers:
        passenger.raw_cohort = strategy.cohort(passenger)
        passenger.random_key = rng.next()
        passenger.raw_rank = strategy.rank(passenger, passenger.random_key)
        passenger.assigned_door = strategy.door(passenger)
        passenger.prep_cohort = passenger.raw_cohort
        if passenger.family_id:
            families.setdefault(passenger.family_id, []).append(passenger)

    for members in families.values():
        cohort = min(member.raw_cohort for member in members)
        front_count = sum(member.assigned_door == "front" for member in members)
        rear_count = len(members) - front_count
        common_door = members[0].assigned_door if front_count == rear_count else ("front" if front_count > rear_count else "rear")
        for member in members:
            original_door = member.assigned_door
            member.prep_cohort = cohort
            if not strategy.preserve_seat_door:
                member.assigned_door = common_door
            member.companion_override = member.raw_cohort != cohort or member.assigned_door != original_door

    groups: list[list[Passenger]] = []
    grouped_ids: set[int] = set()
    for family_id in sorted(families):
        members = sorted(families[family_id], key=lambda passenger: (passenger.raw_rank, passenger.id))
        groups.append(members)
        grouped_ids.update(member.id for member in members)
    groups.extend([[passenger] for passenger in passengers if passenger.id not in grouped_ids])
    groups.sort(key=lambda group: (min(member.raw_rank for member in group), min(member.id for member in group)))
    next_rank = 0
    for group in groups:
        for member in group:
            member.boarding_rank = float(next_rank)
            next_rank += 1
    return passengers

File: src/strategies.js
const seatGroup = seat => (seat === 'A' || seat === 'F') ? 0 : (seat === 'B' || seat === 'E') ? 1 : 2;
const side = seat => ['A','B','C'].includes(seat) ? 0 : 1;
const zoneBackToFront = row => Math.floor((30 - row) / 5);

export const STRATEGIES = {
  random_front: {
    id: 'random_front', name: 'Random · front door', accessRecommended: 'bridge', prepCohorts: 1,
    cohort: () => 0,
    rank: (p, randomKey) => randomKey,
    door: () => 'front'
  },
  split_half_two_door: {
    id: 'split_half_two_door', name: 'Rows 1–15 front · 16–30 rear', accessRecommended: 'bus', prepCohorts: 2, preserveSeatDoor: true,
    cohort: p => p.row <= 15 ? 0 : 1,
    rank: (p, randomKey) => (p.row <= 15 ? 0 : 1) * 1000 + randomKey,
    door: p => p.row <= 15 ? 'front' : 'rear'
  },
  wilma: {
    id: 'wilma', name: 'A/F → B/E → C/D', accessRecommended: 'bridge', prepCohorts: 3,
    cohort: p => seatGroup(p.seat),
    rank: (p, randomKey) => seatGroup(p.seat) * 1000 + randomKey,
    door: () => 'front'
  },
  back_to_front_zones: {
    id: 'back_to_front_zones', name: 'Back-to-front · 5-row zones', accessRecommended: 'bridge', prepCohorts: 6,
    cohort: p => zoneBackToFront(p.row),
    rank: (p, randomKey) => zoneBackToFront(p.row) * 1000 + randomKey,
    door: () => 'front'
  },
  wilma_zones: {
    id: 'wilma_zones', name: 'Outside-in + back-to-front zones', accessRecommended: 'bridge', prepCohorts: 18,
    cohort: p => seatGroup(p.seat) * 6 + zoneBackToFront(p.row),
    rank: (p, randomKey) => (seatGroup(p.seat) * 6 + zoneBackToFront(p.row)) * 1000 + randomKey,
    door: () => 'front'
  },
  steffen_companion: {
    id: 'steffen_companion', name: 'Steffen-style · companion compatible', accessRecommended: 'bridge', prepCohorts: 12,
    cohort: p => seatGroup(p.seat) * 4 + side(p.seat) * 2 + (p.row % 2),
    rank: (p) => {
      const sg = seatGroup(p.seat), sd = side(p.seat), parity = p.row % 2;
      const descendingRow = 31 - p.row;
      return (sg * 4 + sd * 2 + parity) * 1000 + descendingRow;
    },
    door: () => 'front'
  },
  split_wilma_two_door: {
    id: 'split_wilma_two_door', name: 'Split doors + A/F → B/E → C/D', accessRecommended: 'bus', prepCohorts: 6, preserveSeatDoor: true,
    cohort: p => (p.row <= 15 ? 0 : 1) * 3 + seatGroup(p.seat),
    rank: (p, randomKey) => ((p.row <= 15 ? 0 : 1) * 3 + seatGroup(p.seat)) * 1000 + randomKey,
    door: p => p.row <= 15 ? 'front' : 'rear'
  }
};

export function strategyComplexity(strategy) {
  // Structural complexity only: monotonic with the number of distinct cohorts.
  // This is not a measured human-factors coefficient.
  return strategy.prepCohorts <= 1 ? 0 : Math.log2(strategy.prepCohorts) / Math.log2(18);
}

export function applyCompanionCompatibility(passengers, strategy, rng) {
  for (const p of passengers) {
    p.rawCohort = strategy.cohort(p);
    p.randomKey = rng.next();
    p.rawRank = strategy.rank(p, p.randomKey);
    p.assignedDoor = strategy.door(p);
  }
  const families = new Map();
  for (const p of passengers) if (p.familyId) {
    if (!families.has(p.familyId)) families.set(p.familyId, []);
    families.get(p.familyId).push(p);
  }
  for (const members of families.values()) {
    // Companions stay together. They take the earliest cohort/rank represented in the family.
    const cohort = Math.min(...members.map(p => p.rawCohort));
    const rank = Math.min(...members.map(p => p.rawRank));
    // A family cannot be split between aircraft doors. Use the door serving the majority of seats;
    // ties use the leader's door. This may reduce ideal strategy conformance, intentionally.
    const front = members.filter(p => p.assignedDoor === 'front').length;
    const rear = members.length - front;
    const commonDoor = front === rear ? members[0].assignedDoor : front > rear ? 'front' : 'rear';
    for (let k = 0; k < members.length; k++) {
      members[k].prepCohort = cohort;
      members[k].boardingRank = rank + k * 0.001;
      // For two-door half-cabin methods, preserve the seat-based door assignment.
      // Otherwise a family spanning the midpoint would create opposing aisle traffic.
      // The family stays in one preparation cohort but separates only at the aircraft-door branch.
      if (!strategy.preserveSeatDoor) members[k].assignedDoor = commonDoor;
      members[k].companionOverride = members[k].rawCohort !== cohort || (!strategy.preserveSeatDoor && strategy.door(members[k]) !== commonDoor);
    }
  }
  for (const p of passengers) if (!p.familyId) {
    p.prepCohort = p.rawCohort;
    p.boardingRank = p.rawRank;
    p.companionOverride = false;
  }
  return passengers;
}
